contracts/src/StakeTableV2.sol

```diff
@@ -94,32 +101,130 @@ contract StakeTableV2 is StakeTable, PausableUpgradeable, AccessControlUpgradeab
+    /// @param validator The validator to withdraw from
+    /// @dev This function is overridden to deduct the amount from the validator's delegatedAmount
+    /// and to add pausable functionality
+    /// @dev since the delegated Amount is no longer updated during undelegation
+    /// @dev delegatedAmount represents the no. of tokens that have been delegated to a validator,
+    /// even if it's not participating in consensus
     function claimWithdrawal(address validator) public virtual override whenNotPaused {
-        super.claimWithdrawal(validator);
+        address delegator = msg.sender;
+        // If entries are missing at any of the levels of the mapping this will return zero
+        uint256 amount = undelegations[validator][delegator].amount;
+        if (amount == 0) {
+            revert NothingToWithdraw();
+        }
+
+        if (block.timestamp < undelegations[validator][delegator].unlocksAt) {
+            revert PrematureWithdrawal();
+        }
+
+        // Mark funds as spent
+        delete undelegations[validator][delegator];
+        // deduct the amount from the validator's delegatedAmount
+        validators[validator].delegatedAmount -= amount;
+
+        SafeTransferLib.safeTransfer(token, delegator, amount);
+
+        emit Withdrawal(delegator, amount);
     }
 
+    /// @notice Delegate funds to a validator
+    /// @param validator The validator to delegate to
+    /// @param amount The amount to delegate
+    /// @dev This function is overridden to add pausable functionality
     function delegate(address validator, uint256 amount) public virtual override whenNotPaused {
         super.delegate(validator, amount);
     }
 
+    /// @notice Undelegate funds from a validator
+    /// @param validator The validator to undelegate from
+    /// @param amount The amount to undelegate
+    /// @dev This function is overridden to add pausable functionality
+    /// @dev and to ensure that the validator's delegatedAmount is not updated until withdrawal
+    /// @dev delegatedAmount represents the no. of tokens that have been delegated to a validator,
+    /// even if it's not participating in consensus
     function undelegate(address validator, uint256 amount) public virtual override whenNotPaused {
-        super.undelegate(validator, amount);
+        ensureValidatorActive(validator);
+        address delegator = msg.sender;
+
+        if (amount == 0) {
+            revert ZeroAmount();
+        }
+
+        if (undelegations[validator][delegator].amount != 0) {
+            revert UndelegationAlreadyExists();
+        }
+
+        uint256 balance = delegations[validator][delegator];
+        if (balance < amount) {
+            revert InsufficientBalance(balance);
+        }
+
+        delegations[validator][delegator] -= amount;
+        undelegations[validator][delegator] =
+            Undelegation({ amount: amount, unlocksAt: block.timestamp + exitEscrowPeriod });
```

Comment:
Should we really remove decrementing `validator.delegatedAmount` here?

I think it would be better to keep that.

There are 2 ways for funds to be undelegated: undelegation, exit. For exit we can't do the accounting in the contract on exit because it would mean iterating over all delegators so we delay it until funds are claimed by the delegator.

With the current change I think we have the problem that: sum of delegated amounts to validator != validator.delegatedAmounts. So it seems to me like we're trading one problem for another, instead of fixing a problem.

cc @0xkato @alysiahuggins

Reply:
We should probably add whatever invariants we want to have to the invariant tests. So that we have some guarantees that they are satisfied and have the invariants captured more explicitly.

Reply:
thank you @sveitser,i agree. my intentions were to have the validator.delegatedAmount only be adjusted upon withdrawals so that the semantics could be more easily understood but i agree with you and would prefeer that the sum of delegatedAmounts == validator.delegatedAmounts